### PR TITLE
[DependencyInjection] Improve error message when a class in not instantiable during auto-discovery

### LINKED ISSUE #53651: [DI] Getting the wrong error message ("no such service exists, did you create a class that implements this interface?" when class exists but its constructor is private)
### Symfony version(s) affected

6.3.7

### Description

Hello there!

I created an interface and a class implementing it and injected it in another's class constructor. But I made a mistake, and created the implementation with a private constructor instead of a public one.

The framework failed with the following error message : 

> Cannot autowire service "App\Domain\X": argument "$y" of method "__construct()" references interface "App\Domain\YInterface" but no such service exists. Did you create a class that implements this interface?

Therefore, I started to triple-check that the class name, namespace, filename were all correct, tried commenting a bunch of dependencies, rebuilt composer's autoloader, restarted my Docker env... Nothing did the trick.

Then I thought to try to register the service by hand in `config/services.php`, and then the error changed to : 

> Invalid service "App\Domain\YInterface": constructor of class "App\Domain\Y" must be public.

That was it! Changed the constructor from private to public, and it was fixed. It would have been great if I didn't lost half an hour because the error wasn't the correct one though. :)

Hope it could be fixed in the future, though I understand it's low priority.

Have a nice day!
gnutix

### How to reproduce

Inject a class with a private constructor in another one.

### Possible Solution

Display the proper error message.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by HMRDevil**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #53651
| License       | MIT

See #53651

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 1]
```diff
@@ -514,6 +514,7 @@ private function createTypeNotFoundMessage(TypedReference $reference, string $la
 
             if ($r->isInterface() && !$alternatives) {
                 $message .= ' Did you create a class that implements this interface?';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]:
```suggestion
                $message .= ' Did you create an instantiable class that implements this interface?';
```

**Review by nicolas-grekas — COMMENTED** [review]
What about doing just this?
Note that this should target 7.2: it's not a bugfix but an improvement.

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 1 above)
Comment (HMRDevil) [reply src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: @nicolas-grekas Thanks for the comment. But can we be sure that this is the best way?
The purpose of my PR is to provide the user with as much information about the error as possible. This is especially true for beginners.
I would also like to receive comments from other team members, what they think about this.

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: A private constructor is not the only reason: an abstract class is another one that could end up with this message. I'm all for clearer messages, but here this looks like documentation in a message, and it's not complete. The adjective "instantiable" is enough this to hint the brain on the correct path IMHO.

Comment (HMRDevil) [reply src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: Hm, maybe you're right. Thanks for the tip

**Review by nicolas-grekas — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @HMRDevil.

**Pull request merged**